src/goto-instrument/replace_function_bodies.cpp

```diff
@@ -0,0 +1,367 @@
+/*******************************************************************\
+
+Module: Replace bodies of goto functions
+
+Author: DiffBlue Limited.
+
+\*******************************************************************/
+
+#include "replace_function_bodies.h"
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <memory>
+#include <sstream>
+#include <util/make_unique.h>
+#include <util/message.h>
+#include <langapi/language_util.h>
+
+class replace_function_bodiest
+{
+protected:
+  virtual void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) = 0;
+
+public:
+  virtual ~replace_function_bodiest() = default;
+
+  void replace_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name)
+  {
+    function.body.clear();
+    generate_parameter_names(function, symbol_table, function_name);
+    generate_function_body(function, symbol_table, function_name);
+  }
+
+private:
+  /// Generate parameter names for unnamed parameters.
+  /// CBMC expect functions to have parameter names
+  /// if they are called and have a body
+  void generate_parameter_names(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) const
+  {
+    namespacet ns(symbol_table);
+    int param_counter = 0;
+    for(auto &parameter : function.type.parameters())
+    {
+      if(
+        parameter.type().id() == ID_pointer &&
+        !is_constant_or_has_constant_components(parameter.type(), ns))
+      {
+        if(parameter.get_identifier().empty())
+        {
+          const std::string param_base_name =
+            parameter.get_base_name().empty()
+              ? "__param$" + std::to_string(param_counter++)
+              : id2string(parameter.get_base_name());
+          irep_idt new_param_identifier =
+            id2string(function_name) + "::" + param_base_name;
+          parameter.set_base_name(param_base_name);
+          parameter.set_identifier(new_param_identifier);
+          parameter_symbolt new_param_sym;
+          new_param_sym.name = new_param_identifier;
+          new_param_sym.type = parameter.type();
+          new_param_sym.base_name = param_base_name;
+          new_param_sym.is_parameter = true;
+          new_param_sym.mode = ID_C;
+          const symbolt *function_symbol = symbol_table.lookup(function_name);
+          new_param_sym.module = function_symbol->module;
+          new_param_sym.location = function_symbol->location;
+          symbol_table.add(new_param_sym);
+        }
+        auto function_symbol = *symbol_table.lookup(function_name);
```

Comment:
Also what is going on here? Should this not be done after updating all the parameters? Also I think if you want to update the symbol table you should be using `get_writeable` rather than removing and adding an entry

Reply:
Yes it should, this was a braces-derp on my end